cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,213 @@
+configurations don't need to round trip through the Kubernetes API Server.
+This adapter would be an alternative source of endpoints and would not replace
+the existing default behavior of reading directly from Kubernetes APIs.
+
+## Motivation
+
+There are some capabilities that the Kubernetes API Server is not well equipped
+to handle. This includes topology aware routing (or any form of routing that
+requires a feedback loop) and routing to endpoints outside of the cluster.
+
+## Goals
+
+* Introduce a new xDS adapter to Cilium covering ClusterIP routing capabilities.
+* Develop shared interface between Cilium and Kubernetes, KVStore, and xDS as
+  data sources.
+* Provide room for future growth in xDS support beyond ClusterIP routing.
+* Allow the use of any combination of independent data sources (Kubernetes,
+  KVStore, xDS) with Cilium.
+
+## Non-Goals
+
+* Build anything additional on top of this xDS adapter. Although this could be a
+  foundational technology for several new features, those are out of scope for
+  this proposal.
+
+## Background
+
+Several years ago, Matt Klein published a blog post charting a vision for xDS as
+a [universal data plane
+API](https://blog.envoyproxy.io/the-universal-data-plane-api-d15cec7a). Although
+these APIs are primarily used by Envoy and gRPC today, there has been a broader
+goal in the community to enable more widespread usage as part of a truly
+universal data plane API. xDS is CNCF-governed and is gradually transitioning
+content from the
+[envoyproxy/data-plane-api](https://github.com/envoyproxy/data-plane-api) repo
+to [cncf/xds](https://github.com/cncf/xds) repo as part of that overall vision.
+
+In parallel to these efforts by the xDS community, GKE is planning to introduce
+xDS as an additional data source for DPv2 configuration. This feels sufficiently
+generic and helpful that it could be something that could be contributed to
+upstream Cilium. This could be particularly useful for at least two common use
+cases:
+
+1. Supporting Services and Endpoints from outside of the local cluster.
+1. Supporting advanced routing techniques, such as topology aware routing.
+
+Although it is out of scope for this specific CFP to provide complete solutions
+for either of these use cases, it will demonstrate the benefits of having an
+xDS adapter when developing a solution for either of these use cases.
+
+### Use Cases
+
+#### Topology Aware Routing
+
+As the author of both EndpointSlices and the existing Topology Aware Routing
+approach in Kubernetes, I feel like I’m well equipped to discuss the limitations
+of both. What most people want from Topology Aware Routing is essentially to
+fill the local zone to capacity and then spillover to the next closest set of
+endpoints with capacity. Unfortunately that is very difficult to achieve with
+the existing Kubernetes tooling. There are a few key problems here:
+
+1. Kubernetes has no feedback loop to its dataplane and therefore no reliable
```

Comment:
Just to be clear, in Cilium, an "xDS adapter" alone doesn't solve this. We will also need changes into Cilium's control plane to propagate its internal state back to the xDS adapter.

Reply:
+1, this just provides the foundation to solve this. The whole idea of topology aware routing or really any form of routing that requires a feedback loop is deceptively simple until you start to think through this part of it. What xDS gives us here is an established API + patterns for completing a feedback loop, but like you're saying, we'd still need to connect the dots here.